Metrics Validation and Verification (#1169)
kubernetes/enhancements: keps/sig-instrumentation/20190605-metrics-validation-and-verification.md

  - [Should code snippets used for unit testing be kept in separate files instead of strings?](#should-code-snippets-used-for-unit-testing-be-kept-in-separate-files-instead-of-strings)
<!-- /toc -->

## Summary

 This Kubernetes Enhancement Proposal (KEP) builds off of the framework proposed
in the [Kubernetes Control-Plane Metrics Stability
KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20190404-kubernetes-control-plane-metrics-stability.md)
and proposes a strategy for ensuring conformance of metrics with official
stability guarantees.

## Motivation

 While the [Kubernetes Control Plane metrics stability KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20190404-kubernetes-control-plane-metrics-stability.md) provides a framework to define stability levels for control-plane metrics, it does not provide a strategy for verifying and validating conformance to stated guarantees. This KEP intends to propose a framework for validating and verifying metric guarantees.

### Goals

* Given a stable metric, validate that we cannot remove or modify it (other than adding deprecation information).

> * Given a deprecated but stable metric, validate that we cannot remove or modify it until the deprecation period has elapsed.

Reviewer: Where is this deprecation period stated, I haven't seen it mentioned in this doc (could have missed it). I have seen in the original KEP, as `DeprecatedVersion`, is this what this is referring to. If yes, any plans on how will we check in which kubernetes release this metric will land to make sure we can remove it.

Reviewer 2: It might be worth reiterating in this document, but if I recall correctly the version a metric was first deprecated in is part of the metric's metadata.

Author: Yes, DepracatedVersion is part of metadata put in `stable-metrics-list.yaml`. This means that decision of deprecating metric will require approval from sig-instrumentation.

* Given an alpha metric which is promoted to be 'stable', automatically include proper instrumentation reviewers (for schema validation and conformance to metrics guidelines).

### Non-Goals

* Conformance testing will not apply to alpha metrics, since alpha metrics do not have stability guarantees.

## Proposal

We will provide validation for metrics under the [new framework](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20190404-kubernetes-control-plane-metrics-stability.md) with static analysis.

## Design Details

> Metrics conformance testing will work in a similar (but not identical) fashion to the generic Kubernetes conformance tests.  Sig-instrumentation will own a directory under `test/instrumentation`. There will be a subdirectory `testdata` in which a file `stable-metrics-list.txt` will live.  This file will be owned by sig-instrumentation. Metrics conformance tests will involve a static analysis script which will traverse the entire codebase and look for metrics which are annotated as 'stable'. For each stable metrics, this script will generate a stringified version of metric metadata (i.e. name, type, labels) which will then be appended together in lexographic order. This will be the output of this script.

Reviewer 3: This `stable-metrics-list.txt` file also seems like a place that we could include/use to generate per-metric documentation, which could significantly improve discoverability and usability of out-of-the-box k8s metrics.

Author: This would be a useful idea, but it's outside of scope of this KEP.
@brancz @logicalhan WDYT?

Reviewer 2: I’m ok treating it separately, or maybe add it as a stretch goal as the work for it should be almost identical. I’m ok calling this KEP done without it though.

Author: I will add it to stretch goal and update section about list format.
Thanks

> Metrics conformance testing will work in a similar (but not identical) fashion to the generic Kubernetes conformance tests.  Sig-instrumentation will own a directory under `test/instrumentation`. There will be a subdirectory `testdata` in which a file `stable-metrics-list.txt` will live.  This file will be owned by sig-instrumentation. Metrics conformance tests will involve a static analysis script which will traverse the entire codebase and look for metrics which are annotated as 'stable'. For each stable metrics, this script will generate a stringified version of metric metadata (i.e. name, type, labels) which will then be appended together in lexographic order. This will be the output of this script.

Reviewer 2: I'm not sure "Conformance testing" is a good label for this, as that has special meaning in kubernetes, and I'm unsure that's what we really want here.

Author: Changed to "Stable metrics testing"

We will add a pre-submit check, which will run in our CI pipeline, which will run our script with the current changes and compare that to existing, committed file. If there is a difference, the pre-submit check will fail. In order to pass the pre-submit check, the original submitter of the PR will have to run a script `test/instrumentation/update-stable-metrics.sh` which will run our static analysis code and overwrite `stable-metrics-list.txt`. This will cause `sig-instrumentation` to be tagged for approvals on the PR (since they own that file).
 
 
### Static Analysis
 

> Similarly to conformance test proposed analysis will be performed on golang source code using default abstract syntaxt tree parser `go/ast`. Handling all possible cases for how metrics can be instantiated would require executing the code itself and is not practical. To reduce number of cases needed handled we will be making assumption of non malicious intent of contributors. There are just to many ways on golang struct can be instantiated that would hide potential stable metrics. To ensure correctness and code simplicity we will be restricting how stable metrics can be declared to one format described below. Alpha metrics will not restricted in the same way to allow flexibility and their dynamic generation.

Reviewer 3: ```suggestion
There are just too many ways one golang struct can be instantiated that would hide potential stable metrics.
```

(as a side note, I wish this file's lines flowed at 80-char width, it's hard to comment/make suggestions on these massive paragraphs)

Author: Applied your change and reformatted text to your suggestion. Thanks!

> Similarly to conformance test proposed analysis will be performed on golang source code using default abstract syntaxt tree parser `go/ast`. Handling all possible cases for how metrics can be instantiated would require executing the code itself and is not practical. To reduce number of cases needed handled we will be making assumption of non malicious intent of contributors. There are just to many ways on golang struct can be instantiated that would hide potential stable metrics. To ensure correctness and code simplicity we will be restricting how stable metrics can be declared to one format described below. Alpha metrics will not restricted in the same way to allow flexibility and their dynamic generation.

Reviewer 4: We should also note that alpha metrics don't need to be analyzed at all since there are no stability guarantees. That's the driving reason why one could do whatever they wanted with alpha metrics.

Author: Changed last sentence to "Alpha metrics will not be analyzed as they don't have any stability guarantees. Their declaration will not have any restrictions, thus allowing dynamic generation."

Those restrictions will allow AST based analysis to correctly interpret metric definitions.

### Stable metric detection strategy 

Static analysis will be done in two steps. First find stable metrics, second read and validate their fields. 

In first one step we will want to distinguish stable metric without relying on their definition structure, allowing freedom of declaration for non-stable metrics. This will be done by finding occurrences of metric options object initialization (`CounterOpts` for `Counter`). If this initialization sets `StabilityLevel` then it will validate and read it. If value could not be extracted then script should fail. Metrics which set `kubemetrics.STABLE` will be passed to second step.

Second step will be extracting information from stable metric definitions. This step will be expecting exact structure of AST subtree of new metric call matching format proposed above. If metric options object was initialized outside of new metric call or call structure deviates from expected format then static analysis should fail. This restriction will ensure that no field was read incorrectly or stable metric missed.

### Format of stable metrics list file

> This file should include information about metrics needed to detect a breaking change. Those fields should include `name`, `namespace`, `subsystem` as they are used to build metric fully-qualified name, metric type and metric labels. Metric labels should include both vector labels and const labels. 

Reviewer 2: `subsystem` is likely to be removed one day in the prometheus client, it'd be better not to use it

Author: "subsystem" is a part of Prometheus metrics fully qualified name. So we need to include it if is used. If it's not set it will not be in put in `yaml`. We could potentialy remove it when it's removed from  KubeOps for metrics in metric framework

Reviewer 2: There is no way to make a distinction between subsystem and name, hence it's being removed, and we're creating a new API and should go that way right now already.

Metric information stored in file should be in easly readable and reviewable format. For that pretty printed json will be used. Example:

> ```json

Reviewer 3: YAML might be a little more human-readable? Even when pretty-printed, long JSON strings can be frustrating to review.

Reviewer 2: I think I'd prefer YAML as well for reviews

Author: That's great suggestion.

[
  {
    "name": "some_metric",
    "subsystem": "some_subsystem",
    "namespace": "some_namespace",
    "labels": [
      "some-label"
    ],
    "type": "Counter"
  }
]
```

Json keys and lists will be sorted alphabetically to ensure that there is only one correct output.

### Failure Modes

> In cases where static analysis could not be done correctly (invalid stable metric format, etc.) script should fail (non zero response code) and write to STDERR information on problem. Example:

Reviewer 4: In cases where static analysis could not be done correctly [on a stable metric]

Author: There is small restrictions for non-stable metrics,  `StabilityLevel` cannot be auto-generated as we need it to recognize stable ones.

```
pkg/controller/volume/persistentvolume/scheduler_bind_cache_metrics.go:42 stable metric CounterOpts should be invoked from newCounter
pkg/controller/volume/persistentvolume/scheduler_bind_cache_metrics.go:44 stable metric CounterOpts should be invoked from newCounter
exit status 1
```
 
 ### Performance
 
Proposed static analysis takes around 4 seconds on one CPU core. This time is comparable to conformance tests check ~2 sec and should be acceptable.

 ## Graduation Criteria

- [ ] Stable metric validation is enabled as pre-submit check

 ## Implementation History

- [ ] Implement static analysis script analysis all metrics types
 
 ## Test Plan
 
- [ ] Static analysis is covered by unit tests using common patterns of metric definition
 
 ## Unresolved questions
 

>  ### Should changes in histogram buckets/summary objectives be a breaking change?

Reviewer 4: Label value changes should be ingestion compatible.

Author: Discussion in other thread.

 
Histogram buckets are usually used to calculate percentile and not directly used. This means that adjusting of buckets will not directly break their users as they should not hardcode bucket values.

> Story can be somewhat reversed for Summary objectives. Users will be directly picking set of preferred quantiles (p99) and hardcoding their values in queries.

Reviewer 3: I don't understand this comment. What does "story can be somewhat reversed for Summary" mean? That we do need to include stability guarantees for Summary bucketing?

Reviewer 2: Agreed

Author: It would be useful to discuss what we want to treat as breaking for purpose of this KEP.
Based on @logicalhan other comment 
`Label value changes should be ingestion compatible` options like Summary objective would not be breaking as it would only change only label values. Still in my opinion it would impact user as described.
@brancz WHYT? 
Deciding on which fields are breaking could be postponed if we decided to store all metadata in `stable-metrics-list.yaml`. It would be up to reviewer to decide.

Reviewer 2: Agreed. I was actually thinking this further and I think it could be argued that a new label could actually be backward compatible. Actually all aggregations would continue to work, only where single time-series are expected there could be multiple, which in a vector based system is an odd expectation so I'm thinking adding a new label would actually be ok.

Author: I went with approach of leaving it up to reviewer to decide if change to metric is backward compatible. This requires adding all metric metadata, but as result we will get stretch goal proposed by @ehashman of generating source of metric documentation. 

@logicalhan WDYT? We wanted only to monitor backward incompatible changes but it will be hard to imminently pin down which fields we should include.

Reviewer 4: Adding new labels is not backwards compatible because depending on your ingestion pipeline, it will not work out of the box. Adding new label values will work because the schema of your metric does not change.

Reviewer 4: Just because new labels are possibly backwards compatible doesn't mean that they are backwards compatible. We should only be considering worst cases here, since we are looking to avoid breakages.

Reviewer 2: I'm saying let's not categorically say no, I think we should decide when we actually have such a case and can analyze its effect appropriately.

---
title: Metrics Validation and Verification
authors:
  - "@serathius"
  - "@solodov"
  - "@logicalhan"
owning-sig: sig-instrumentation
participating-sigs:
  - sig-instrumentation

>   - sig-testing

Reviewer 5: Who from sig testing is participating in this? Has this been reviewed by sig-testing at all?
cc: @brancz @serathius @solodov @logicalhan

 To ensure correctness and code simplicity we will be restricting how stable metrics can be declared to one format described below.
 Alpha metrics will not be analyzed as they don't have any stability guarantees.
 Their declaration will not have any restrictions, thus allowing dynamic generation.

### Discovery of golang source files

Stable metric list will be a bazel genrule provided with locations of files from `//:all-src` target.
This target is auto-managed, validate in CI by `./hack/validate-bazel.sh` and should include all files in repository.
 Golang source files will be filtered by extension.

List of skipped directories:
* `vendor/` - Kubernetes metrics from external repos will be shared metrics that will be defined in k/k and injected as
              dependency during registration

### Format of defining stable metrics

To simplify static analysis implementation and reduce chance of missing metrics, we will restrict how stable metrics can be defined.
Stable metrics will use exactly the same functions as normal ones, but code defining them will need to comply to the specific format:

> * Metric options (`CounterOpts`) needs to be directly passed to function (`kubemetrics.NewCounterVec`).

Reviewer 2: These are just examples so we should makes sure to mark them as such in this text

Author: Added "e.g.". Thanks

Metrics which set `kubemetrics.STABLE` will be passed to second step.

Second step will be extracting information from stable metric definitions.
This step will be expecting exact structure of AST subtree of new metric call matching format proposed above.
If metric options object was initialized outside of new metric call or call structure deviates from expected format then
static analysis should fail. This restriction will ensure that no field was read incorrectly or stable metric missed.

### Format of stable metrics list file

This file will include all metric metadata that will serve source of documentation and a way to detect breaking change.
Every change to metric definition will require updating this file, thus a review from sig-instrumentation member.
It will be up to reviewer to decide which changes are backward compatible and which not.

Metric information stored in file should be in easly readable and reviewable format.
For that `yaml` will be used. Example:
```yaml
- deprecatedVersion: "1.16"
  help: "some help information about the metric"
  labels:
    - some-label
  name: some_metric
  namespace: some_namespace
  stabilityLevel: STABLE

>   objectives:

Reviewer 2: can you explain this field more? it seems related to the summary type?

Author: Yes, I change the example metric type to summary to show that all metadata arguments passed to metric and opts will be in `stable-metric-list.yaml`
